### PR TITLE
[3.4] precondition plan / version in compaction / store TTL removal 

### PR DESCRIPTION
**Pull request by kvahed**

### CODE REVIEW

File: arangod/Agency/Store.cpp [hunk 1]
```diff
@@ -135,6 +136,7 @@ Store& Store::operator=(Store const& rhs) {
 Store& Store::operator=(Store&& rhs) {
   if (&rhs != this) {
     MUTEX_LOCKER(otherLock, rhs._storeLock);
+    MUTEX_LOCKER(lock, _storeLock);
```
Comment (maierlars) [inline arangod/Agency/Store.cpp]: was this missing in general or was there a change such that this lock becomes necessary. If the later is true, where is the corresponding lock?

Comment (kvahed) [reply arangod/Agency/Store.cpp]: This was missing as far as I can tell.

File: arangod/Agency/AgencyComm.cpp [hunk 2]
```diff
@@ -1797,10 +1797,9 @@ bool AgencyComm::tryInitializeStructure() {
     LOG_TOPIC(TRACE, Logger::AGENCYCOMM)
         << "Initializing agency with " << builder.toJson();
 
-    AgencyOperation initOperation("", AgencyValueOperationType::SET, builder.slice());
-
-    AgencyWriteTransaction initTransaction;
-    initTransaction.operations.push_back(initOperation);
+    AgencyWriteTransaction initTransaction(
+      AgencyOperation("", AgencyValueOperationType::SET, builder.slice()),
+      AgencyPrecondition("Plan", AgencyPrecondition::Type::EMPTY, true));
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/AgencyComm.cpp]: I intentionally in put only the sacrilege of `Plan` in there. Without the Plan everything else is just rubbish. With

**Comment by fceller** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3653/

File: arangod/Agency/AgencyComm.cpp [hunk 3]
```diff
@@ -1821,19 +1820,43 @@ bool AgencyComm::tryInitializeStructure() {
 }
 
 bool AgencyComm::shouldInitializeStructure() {
-  VPackBuilder builder;
-  builder.add(VPackValue(false));
 
-  // "InitDone" key should not previously exist
-  auto result = casValue("InitDone", builder.slice(), false, 60.0,
-                         AgencyCommManager::CONNECTION_OPTIONS._requestTimeout);
+  while (!application_features::ApplicationServer::isStopping()) {
```
Comment (neunhoef) [inline arangod/Agency/AgencyComm.cpp]: Why another loop here?

File: arangod/Agency/AgencyComm.cpp [hunk 4]
```diff
@@ -1821,19 +1820,43 @@ bool AgencyComm::tryInitializeStructure() {
...
+        if (result.slice()[0] == VPackSlice::emptyObjectSlice()) {
+          LOG_TOPIC(DEBUG, Logger::AGENCYCOMM)
+            << "agency initialisation should be performed";
+          return true;
+        } else {
+          LOG_TOPIC(DEBUG, Logger::AGENCYCOMM)
+            << "agency initialisation under way or done";
+        }
+      } else {
+        // Should never get here
```
Comment (neunhoef) [inline arangod/Agency/AgencyComm.cpp]: Should this not be a FATAL error with a proper log message?

File: arangod/Agency/Store.cpp (hunk 1 above)
Comment (neunhoef) [reply arangod/Agency/Store.cpp]: Have you checked that we do not hold the lock at the caller's site?

File: arangod/Agency/Store.cpp [hunk 5]
```diff
@@ -688,25 +690,29 @@ void Store::clear() {
...
 
-  TRI_ASSERT(slice[1].isObject());
-  for (auto const& entry : VPackObjectIterator(slice[1])) {
-    if (entry.value.isNumber()) {
-      auto const& key = entry.key.copyString();
-      if (_node.has(key)) {
-        auto tp = TimePoint(std::chrono::seconds(entry.value.getNumber<int>()));
-        _node(key).timeToLive(tp);
-        _timeTable.emplace(std::pair<TimePoint, std::string>(tp, key));
+  if (s.hasKey("version")) {
```
Comment (neunhoef) [inline arangod/Agency/Store.cpp]: Please add a comment what we are doing here. We will all have forgotten in 2 weeks time.

File: arangod/Agency/AgencyComm.cpp [hunk 6]
```diff
@@ -1181,23 +1181,23 @@ AgencyCommResult AgencyComm::sendTransactionWithFailover(AgencyTransaction const
...
-      LOG_TOPIC(TRACE, Logger::AGENCYCOMM)
-          << "Agency is fresh. Needs initial structure.";
-      // mop: we are the chosen one .. great success
-
-      if (tryInitializeStructure()) {
-        LOG_TOPIC(TRACE, Logger::AGENCYCOMM)
-            << "Successfully initialized agency";
-        break;
-      }
+  auto serverFeature =
```
Comment (neunhoef) [inline arangod/Agency/AgencyComm.cpp]: We are not allowed to use this here.

File: arangod/Agency/AgencyComm.cpp [hunk 7]
```diff
@@ -1181,23 +1181,23 @@ AgencyCommResult AgencyComm::sendTransactionWithFailover(AgencyTransaction const
...
+        << "Successfully initialized agency";
+      break;
     }
+    
+    LOG_TOPIC(WARN, Logger::AGENCYCOMM)
+      << "Initializing agency failed. We'll try again soon";
+    // We should really have exclusive access, here, this is strange!
+    std::this_thread::sleep_for(std::chrono::seconds(1));
 
     AgencyCommResult result = getValues("InitDone");
```
Comment (neunhoef) [inline arangod/Agency/AgencyComm.cpp]: Why this check here? It is checked in the next loop anyway, as far as I see.

File: arangod/Agency/AgencyComm.cpp (hunk 3 above)
Comment (maierlars) [reply arangod/Agency/AgencyComm.cpp]: Because the agency might not be ready

File: arangod/Agency/AgencyComm.cpp (hunk 4 above)
Comment (kvahed) [reply arangod/Agency/AgencyComm.cpp]: Changed somewhat

File: arangod/Agency/Store.cpp (hunk 5 above)
Comment (kvahed) [reply arangod/Agency/Store.cpp]: Of course

File: arangod/Agency/AgencyComm.cpp (hunk 6 above)
Comment (kvahed) [reply arangod/Agency/AgencyComm.cpp]: OK. Do you have a suggestion how to fix this properly?

File: arangod/Agency/AgencyComm.cpp (hunk 7 above)
Comment (kvahed) [reply arangod/Agency/AgencyComm.cpp]: left over wan't just done yet

File: arangod/Agency/Store.cpp (hunk 1 above)
Comment (kvahed) [reply arangod/Agency/Store.cpp]: I have ran tests don't see anywhere.

**Pull request merged**